fairworkflows/fairstep.py

```diff
@@ -66,24 +68,52 @@ def load_from_nanopub(self, uri):
             if len(concepts_introduced) == 0:
                 raise ValueError('This nanopub does not introduce any concepts. Please provide URI to the step itself (not just the nanopub).')
             elif len(concepts_introduced) > 0:
-                step_uri = concepts_introduced[0]
+                step_uri = str(concepts_introduced[0])
 
             print('Assuming step URI is', step_uri)
 
         else:
             step_uri = uri
 
         self._uri = step_uri
-        self.self_ref = rdflib.URIRef(self._uri)
 
         # Check that the nanopub's assertion actually contains triples refering to the given step's uri
-        if (rdflib.URIRef(self.self_ref), None, None) not in np.assertion:
+        if (rdflib.URIRef(self._uri), None, None) not in np.assertion:
             raise ValueError(f'No triples pertaining to the specified step (uri={step_uri}) were found in the assertion graph of the corresponding nanopublication (uri={np_uri})')
 
         # Else extract all triples in the assertion into the rdf graph for this step
         self._rdf = rdflib.Graph()
         self._rdf += np.assertion
 
+        # Record that this RDF originates from a published source
+        self._is_published = True
+
+
+    def publish_as_nanopub(self):
+        """
+        Publishes the rdf for this FairStep as a nanopublication.
+        Returns True if published successfully.
+        """
+
+        # If this step has been modified from a previously publised step, include this in the derived_from PROV (if applicable)
+        derived_from = None
+        if self._is_published is True:
+            if self.is_modified is True:
+                derived_from = self._uri
+            else:
+                warnings.warn(f'Cannot publish() FairStep. This step is already published (at {self._uri}) and has not been modified.')
+                return False
+
+        # Publish the rdf of this step as a nanopub
+        np_uri = Nanopub.publish(self._rdf, introduces_concept=self.self_ref, derived_from=derived_from)
+
+        # Set the new (published) URI of this fair step, which should be the nanopub URI plus a fragment given by the name of self.self_ref
+        self._uri = np_uri + '#' + str(self.self_ref)
+        
+        self._is_published = True
+        self._is_modified = False
+
+        return True
```

Comment:
I still think we can just return nothing. Not raising errors is enough indication that the function call succeeded.

Reply:
OK I agree. I've fixed this now, and changed the tests so that they instead check to see if the URI has changed (to assess whether the step has been 'published' or not)